File: solutions/python/sliding-window/max-consecutive-ones-iii.py
# LeetCode 1004 — Max Consecutive Ones III (Medium)
# Category: Sliding Window · Approach: Sliding Window
# Time: O(n) | Space: O(1)
# Source: https://leetcode.com/problems/max-consecutive-ones-iii/

def longest_ones(nums: list[int], k: int) -> int:
    left = zeros = res = 0
    for right, v in enumerate(nums):
        if v == 0:
            zeros += 1
        while zeros > k:
            if nums[left] == 0:
                zeros -= 1
            left += 1
        res = max(res, right - left + 1)
    return res

File: solutions/typescript/sliding-window/max-consecutive-ones-iii.ts
// LeetCode 1004 — Max Consecutive Ones III (Medium)
// Category: Sliding Window · Approach: Sliding Window
// Time: O(n) | Space: O(1)
// Source: https://leetcode.com/problems/max-consecutive-ones-iii/

function longestOnes(nums: number[], k: number): number {
  let left = 0, zeros = 0, res = 0;
  for (let right = 0; right < nums.length; right++) {
    if (nums[right] === 0) zeros++;
    while (zeros > k) {
      if (nums[left] === 0) zeros--;
      left++;
    }
    res = Math.max(res, right - left + 1);
  }
  return res;
}
